packages/go_router/lib/src/builder.dart

```diff
@@ -47,164 +47,303 @@ class RouteBuilder {
+            context, matchList, i + 1, pop, routerNeglect, newParams,
+            navigatorKey: route.navigatorKey);
+        final Widget child = result.widget;
+
+        final Page<dynamic> page =
+            buildPageForRoute(context, state, match, child: child);
+
+        // If this ShellRoute is for a different Navigator, add it to the
+        // pagesForOutOfScopeNavigator list instead.
+        if (route.parentNavigatorKey != null) {
+          pagesForOutOfScopeNavigator[route.parentNavigatorKey]!.add(page);
+        } else {
+          pages.add(page);
+        }
+
+        // If any descendent GoRoutes have pages that are out of scope for this
+        // navigator, Add them to the list of routes that are out of scope. If
+        // they are in scope, add them to the list of pages for this navigator.
+        final Map<GlobalKey<NavigatorState>, List<Page<dynamic>>>
+            pagesOutOfScopeForChildNavigator =
+            result.pagesForOutOfScopeNavigator;
+
+        if (pagesOutOfScopeForChildNavigator.containsKey(route.navigatorKey) &&
+            pagesOutOfScopeForChildNavigator[route.navigatorKey]!.isNotEmpty) {
+          pages.addAll(pagesOutOfScopeForChildNavigator[route.navigatorKey]!);
+        }
+
+        // Include any out of scope pages in _RecursiveBuildResult
+        pagesForOutOfScopeNavigator.addAll(pagesOutOfScopeForChildNavigator);
+
+        i = result.newIndex;
+      }
     }
 
-    // wrap the returned Navigator to enable GoRouter.of(context).go()
-    final Uri uri = Uri.parse(location);
-    return builderWithNav(
-      context,
-      GoRouterState(
-        configuration,
-        location: location,
-        // no name available at the top level
-        name: null,
-        // trim the query params off the subloc to match route.redirect
-        subloc: uri.path,
-        // pass along the query params 'cuz that's all we have right now
-        queryParams: uri.queryParameters,
-        // pass along the error, if there is one
-        error: error,
-      ),
-      Navigator(
-        restorationScopeId: restorationScopeId,
+    // Add any pages that were out of scope to this Navigator if the keys match.
+    if (pagesForOutOfScopeNavigator.containsKey(navigatorKey)) {
+      pagesForOutOfScopeNavigator[navigatorKey]
+          ?.forEach((Page<dynamic> outOfScopePage) {
+        pages.add(outOfScopePage);
+      });
+    }
+
+    Widget? child;
+    if (pages.isNotEmpty) {
+      child = buildNavigator(
+        context,
+        Uri.parse(matchList.location.toString()),
+        matchList.isError ? matchList.error : null,
+        pop,
+        pages,
+        root: startIndex == 0,
+        navigatorKey: navigatorKey,
+      );
+    } else if (startIndex == 0) {
+      // It's an error to have an empty pages list on the root Navigator.
+      throw RouteBuilderError('No pages built for root Navigator');
+    }
+
+    return _RecursiveBuildResult(child ?? const SizedBox.shrink(), pages,
```

Comment:
This should probably be a _NullWidget that throw on build, we have a lot of example in flutter/flutter